AK/AsyncBitStream.h

```diff
@@ -0,0 +1,326 @@
+
+    WordType peek_bits_possibly_past_end() const
+    {
+        return m_current_and_next_word >> m_bit_position;
+    }
+
+    template<typename T = WordType>
+    ErrorOr<T> read_bits(u8 count)
+    {
+        // FIXME: Teach read_bits to read more than 32 bits. This arbitrary limit is present only
+        //        because of performance: we need to have current and next word in a single
+        //        variable (m_current_and_next_word) and if I store them in unsigned __int128
+        //        instead of u64, performance drops dramatically.
+        static_assert(sizeof(T) <= sizeof(WordType));
+        VERIFY(count <= sizeof(T) * 8);
+
+        if (bits_left() < count)
+            return Error::from_errno(EAGAIN);
+
+        T result = peek_bits_possibly_past_end() & ((1ULL << count) - 1);
+        advance_read_head(count);
+        return result;
+    }
+
+    ErrorOr<bool> read_bit()
+    {
+        if (!bits_left())
+            return Error::from_errno(EAGAIN);
+        bool result = m_current_and_next_word >> m_bit_position & 1;
+        advance_read_head(1);
+        return result;
+    }
+
+    void consume_bits(size_t count)
+    {
+        m_bits_consumed += count;
+    }
+
+    template<typename Func>
+    auto rollback_group(Func&& func)
+    {
+        auto bits_left_originally = m_bits_left;
+        auto result = func();
+        if (!result.is_error())
+            consume_bits(bits_left_originally - m_bits_left);
+        return result;
+    }
+
+private:
+    void refill_next_word()
+    {
+        if (!m_aligned_words.is_empty()) {
+            m_current_and_next_word |= static_cast<DoubleWordType>(m_aligned_words[0]) << bits_in_word;
+            m_aligned_words = m_aligned_words.slice(1);
+        } else {
+            m_current_and_next_word |= static_cast<DoubleWordType>(m_unaligned_end) << bits_in_word;
+            m_unaligned_end = 0;
+        }
+    }
+
+    void advance_read_head(u8 bits)
+    {
+        m_bit_position += bits;
+        m_bits_left -= bits;
+        if (m_bit_position >= bits_in_word) {
+            m_bit_position -= bits_in_word;
+            m_current_and_next_word >>= bits_in_word;
+            refill_next_word();
+        }
+    }
+
+    u8 m_bit_position { 0 }; // bit offset inside current word
+    DoubleWordType m_current_and_next_word { 0 };
+    size_t m_bits_left { 0 };
+    size_t m_bits_consumed { 0 };
+
+    ReadonlySpan<WordType> m_aligned_words;
+    WordType m_unaligned_end { 0 };
+};
+
+class AsyncInputLittleEndianBitStream final : public AsyncInputStream {
+    AK_MAKE_NONCOPYABLE(AsyncInputLittleEndianBitStream);
+    AK_MAKE_NONMOVABLE(AsyncInputLittleEndianBitStream);
+
+public:
+    AsyncInputLittleEndianBitStream(MaybeOwned<AsyncInputStream>&& stream)
+        : m_stream(move(stream))
+    {
+    }
+
+    ~AsyncInputLittleEndianBitStream()
+    {
+        if (is_open())
+            reset();
+    }
+
+    void reset() override
```

Comment:
I think we usually fully declare overrides, ie
```suggestion
    virtual void reset() override
```